### PR TITLE
[expo-constants] add getAppConfig scripts

### PR DESCRIPTION
**Pull request by esamelson**
# Why

Support `Constants.manifest` in the bare workflow!

# How

- Created scripts that use the method from https://github.com/expo/expo-cli/pull/2863 to grab the public app config and embed it into a build as `app.config` (alongside updates' `app.manifest` and `app.bundle`) files.
- Copied most of the platform-specific scripts from expo-updates. (The meat of this PR is the `getAppConfig.js` script.)

# Alternative

Rather than embedding the app config and reading/exporting it natively, we could instead do this all on the JS side (e.g. before the RN bundle step, fill in some template JS file that expo-constants imports). I opted for the native path because it's similar to what we already do with expo-updates, and the order of the build phases doesn't matter. But definitely open to changing this.

# Test Plan

Tested the script manually to ensure the resulting file has the desired output.

# Todos

- update `@expo/config` version in this PR after https://github.com/expo/expo-cli/pull/2863 lands and is released
- PR for expo-constants to read manifest from this generated file at runtime (https://github.com/expo/expo/pull/10949)
- PR for expo-constants to read manifest from Updates.manifest & prefer it, if it exists (https://github.com/expo/expo/pull/10668 is part of the way, but it will not have the behavior we want if Updates.manifest is `{}`)
- Figure out what should actually call these scripts, and document them

### CODE REVIEW

**Review by bbarthec — APPROVED** [review]
Looks okay to me 👍
So the idea is to have common code in js, but specific code in platform specific scripts? Android, I'm looking at you 👀

File: packages/expo-constants/scripts/get-app-config-android.gradle [hunk 1]
```diff
@@ -0,0 +1,67 @@
+// Gradle script for downloading assets that make up an OTA update and bundling them into the APK
```
Comment (brentvatne) [inline packages/expo-constants/scripts/get-app-config-android.gradle:1]:
```suggestion
// Gradle script for generating serialized public app config (from app.config.js/ts or app.json) and bundling it into the APK 
```

File: packages/expo-constants/scripts/get-app-config-android.gradle [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      // Create dirs if they are not there (e.g. the "clean" task just ran)
+      doFirst {
+        assetsDir.mkdirs()
+      }
+
+      // Set up inputs and outputs so gradle can cache the result
+      inputs.files fileTree(dir: projectRoot, excludes: inputExcludes)
+      outputs.dir assetsDir
+
+      // Set up the call to exp
```
Comment (brentvatne) [inline packages/expo-constants/scripts/get-app-config-android.gradle:50]:
```suggestion
      // Switch to project root and generate the app config
```

not totally sure if this is useful but i think the previous comment didn't apply here

File: packages/expo-constants/scripts/get-app-config-ios.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+DEST="$CONFIGURATION_BUILD_DIR/$UNLOCALIZED_RESOURCES_FOLDER_PATH"
+NODE_BINARY=${NODE_BINARY:-node}
+
+# Related to: https://github.com/facebook/react-native/blob/c9f869f9c7c8b035a669980382af4bbd4afecb89/scripts/react-native-xcode.sh#L59-L69
+PROJECT_ROOT=${PROJECT_ROOT:-$PWD}
+cd "$PROJECT_ROOT" || exit
+
+if ! [ -x "$(command -v $NODE_BINARY)" ]; then
+  echo 'Error: cannot find the node binary. Try setting the NODE_BINARY variable in the ' \
+  '"Bundle React Native code and images" Build Phase to the absolute path to your node binary. ' \
```
Comment (brentvatne) [inline packages/expo-constants/scripts/get-app-config-ios.sh:14]: i think this should probably use the name of whatever we recommend this build phase be called

Comment (esamelson) [reply packages/expo-constants/scripts/get-app-config-ios.sh:14]: For expo-updates we just use the existing build phase and append a call to this script to the end. We could probably do the same with this one, unless you think it makes sense to create a separate build phase.

Comment (brentvatne) [reply packages/expo-constants/scripts/get-app-config-ios.sh:14]: ah ok, makes sense

**Review by brentvatne — APPROVED** [review]

**Pull request merged**